# Project Sales Analysis
### What has been done:
In this project, the following has been done:

1. Obtain the retail data for an online e-commerce store (Part I)
2. Perform exploratory data analysis (Part II)
3. Analyze the data to get customer insights (Part III)
4. Segment the customer base using RFM method  (Part IV)
5. Perform unsupervised machine learning to cluster customers (Part V)

# Introduction

In this notebook, the following has been done:
1. Import pandas library
2. Download the retail data Excel file 
3. Read the Excel file's sheets
4. Combine the separate sheets into a single DataFrame
5. Export the DataFrame as a CSV file 

### Step 1: Import the following library
- pandas as pd

In [1]:
# Step 1: import the library
import pandas as pd

### Step 2: Download data from UCI
Head on to the <a href="https://archive.ics.uci.edu/ml/datasets/Online+Retail+II">UCI Machine Learning Repository</a> and download the dataset from <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/00502/online_retail_II.xlsx">here</a>. 

### Step 3: Read Excel file into a DataFrame
There are two ways to proceed from here:
1. Store the two sheets in two separate variables first, then combine them into one
2. Do 1 but in a single line of code

In [2]:
# Step 3a: Read Excel sheets into two separate DataFrames
xls = pd.ExcelFile('online_retail_II.xlsx')
df1 = pd.read_excel(xls, 'Year 2009-2010')
df2 = pd.read_excel(xls, 'Year 2010-2011')

In [3]:
# Step 3b: Concatenate two DataFrames into a single DataFrame and reset the index
frames = [df1, df2]
  
df_concat = pd.concat(frames)
display(df_concat)

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
541905  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                              POSTAGE         1   

               InvoiceDate  Price  Customer ID         Country  
0      2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1      2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2      2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3      2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4      2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                    ...    ...          ...             ...  
541905 2011-12-09 12:50:00   2.10      12680.0          France  
541906 2011-12-09 12:50:00   4.15      12680.0          France  
541907 2011-12-09 12:50:00   4.15      12680.0          France  
541908 2011-12-09 12:50:00   4.95      12680.0          France  
541909 2011-12-09 12:50:00  18.00      12680.0          France  

[1067371 rows x 8 columns]

### Step 4: Set 'InvoiceDate' as DataFrame's index
Used 'InvoiceDate' as the index

In [4]:
# Step 4: Set InvoiceDate as index
df_inv = df_concat.set_index('InvoiceDate')

### Step 5: Take a quick look at numerical data using .describe()
Let's see if there is anything right/wrong with the data. Use .describe to look the summary of the data.

In [5]:
# Step 5: Use the .describe method on the DataFrame
data = df_inv.copy()
data.describe()

Quantity         Price    Customer ID
count  1.067371e+06  1.067371e+06  824364.000000
mean   9.938898e+00  4.649388e+00   15324.638504
std    1.727058e+02  1.235531e+02    1697.464450
min   -8.099500e+04 -5.359436e+04   12346.000000
25%    1.000000e+00  1.250000e+00   13975.000000
50%    3.000000e+00  2.100000e+00   15255.000000
75%    1.000000e+01  4.150000e+00   16797.000000
max    8.099500e+04  3.897000e+04   18287.000000

### Step 6: Check rows that contain negative quantity
The minimum is not 0. It's a negative value, and a huge one at that. Not possible!

Subset the DataFrame to get rows that contain only values that are less than 0. 

In [6]:
# Step 6a: Filter rows based on "Quantity" that's less than 0
data.loc[data['Quantity'] < 0]

Invoice StockCode                       Description  \
InvoiceDate                                                                
2009-12-01 10:33:00  C489449     22087          PAPER BUNTING WHITE LACE   
2009-12-01 10:33:00  C489449    85206A      CREAM FELT EASTER EGG BASKET   
2009-12-01 10:33:00  C489449     21895     POTTING SHED SOW 'N' GROW SET   
2009-12-01 10:33:00  C489449     21896                POTTING SHED TWINE   
2009-12-01 10:33:00  C489449     22083        PAPER CHAIN KIT RETRO SPOT   
...                      ...       ...                               ...   
2011-12-09 09:57:00  C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL   
2011-12-09 10:28:00  C581499         M                            Manual   
2011-12-09 11:57:00  C581568     21258        VICTORIAN SEWING BOX LARGE   
2011-12-09 11:58:00  C581569     84978  HANGING HEART JAR T-LIGHT HOLDER   
2011-12-09 11:58:00  C581569     20979     36 PENCILS TUBE RED RETROSPOT   

                     Quantity   Price  Customer ID         Country  
InvoiceDate                                                         
2009-12-01 10:33:00       -12    2.95      16321.0       Australia  
2009-12-01 10:33:00        -6    1.65      16321.0       Australia  
2009-12-01 10:33:00        -4    4.25      16321.0       Australia  
2009-12-01 10:33:00        -6    2.10      16321.0       Australia  
2009-12-01 10:33:00       -12    2.95      16321.0       Australia  
...                       ...     ...          ...             ...  
2011-12-09 09:57:00       -11    0.83      14397.0  United Kingdom  
2011-12-09 10:28:00        -1  224.69      15498.0  United Kingdom  
2011-12-09 11:57:00        -5   10.95      15311.0  United Kingdom  
2011-12-09 11:58:00        -1    1.25      17315.0  United Kingdom  
2011-12-09 11:58:00        -5    1.25      17315.0  United Kingdom  

[22950 rows x 7 columns]

In [7]:
# Step 6b: Filter rows based on "Price" that's less than 0
data.loc[data['Price'] < 0]

Invoice StockCode      Description  Quantity     Price  \
InvoiceDate                                                                   
2010-04-29 13:36:00  A506401         B  Adjust bad debt         1 -53594.36   
2010-07-19 11:24:00  A516228         B  Adjust bad debt         1 -44031.79   
2010-10-20 12:04:00  A528059         B  Adjust bad debt         1 -38925.87   
2011-08-12 14:51:00  A563186         B  Adjust bad debt         1 -11062.06   
2011-08-12 14:52:00  A563187         B  Adjust bad debt         1 -11062.06   

                     Customer ID         Country  
InvoiceDate                                       
2010-04-29 13:36:00          NaN  United Kingdom  
2010-07-19 11:24:00          NaN  United Kingdom  
2010-10-20 12:04:00          NaN  United Kingdom  
2011-08-12 14:51:00          NaN  United Kingdom  
2011-08-12 14:52:00          NaN  United Kingdom

### Step 8: Remove negative quantities of Quantity and Price
There are 22,950 rows that contain negative Quantity, and 5 rows that contain negative Price. 

Remove those to have only positive numbers in the DataFrame. 

Expect a final DataFrame with <strong>1,041,671 rows</strong> with 7 columns

In [8]:
# Step 8: Remove rows that contain negative Quantity and Price
data = data.loc[(data["Quantity"] > 0) & (data["Price"] > 0)]
data

Invoice StockCode                          Description  \
InvoiceDate                                                                  
2009-12-01 07:45:00  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS   
2009-12-01 07:45:00  489434    79323P                   PINK CHERRY LIGHTS   
2009-12-01 07:45:00  489434    79323W                  WHITE CHERRY LIGHTS   
2009-12-01 07:45:00  489434     22041         RECORD FRAME 7" SINGLE SIZE    
2009-12-01 07:45:00  489434     21232       STRAWBERRY CERAMIC TRINKET BOX   
...                     ...       ...                                  ...   
2011-12-09 12:50:00  581587     22899         CHILDREN'S APRON DOLLY GIRL    
2011-12-09 12:50:00  581587     23254        CHILDRENS CUTLERY DOLLY GIRL    
2011-12-09 12:50:00  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE   
2011-12-09 12:50:00  581587     22138        BAKING SET 9 PIECE RETROSPOT    
2011-12-09 12:50:00  581587      POST                              POSTAGE   

                     Quantity  Price  Customer ID         Country  
InvoiceDate                                                        
2009-12-01 07:45:00        12   6.95      13085.0  United Kingdom  
2009-12-01 07:45:00        12   6.75      13085.0  United Kingdom  
2009-12-01 07:45:00        12   6.75      13085.0  United Kingdom  
2009-12-01 07:45:00        48   2.10      13085.0  United Kingdom  
2009-12-01 07:45:00        24   1.25      13085.0  United Kingdom  
...                       ...    ...          ...             ...  
2011-12-09 12:50:00         6   2.10      12680.0          France  
2011-12-09 12:50:00         4   4.15      12680.0          France  
2011-12-09 12:50:00         4   4.15      12680.0          France  
2011-12-09 12:50:00         3   4.95      12680.0          France  
2011-12-09 12:50:00         1  18.00      12680.0          France  

[1041671 rows x 7 columns]

### Step 9: Export the DataFrame as CSV
Removed 22k rows

Export the DataFrame as a CSV for Exploratory Data Analysis in Part II.

In [ ]:
# Step 9: Export the DataFrame as CSV
data.to_csv ('DataFrame.csv')